core/domain/exp_domain.py

```diff
@@ -4149,13 +4149,24 @@ def create_mapping_for_new_version(
         for (state_id, state_name) in state_ids_to_names.iteritems():
             if state_name == 'END' and state_name not in new_exploration.states:
                 # If previous version of exploration has END state but new
-                # version does not have END state. In addition, there is no
-                # commit command corresponding to rename or removal of END
-                # state then END state was added to previouos exploratio
-                # verison by states dict schema migration from v2 to v3 and it
-                # was not added to new exploration version by the migration as
-                # new version does not have any rules having 'END' state as
-                # their destination state name.
+                # version does not have END state then exception will be raised
+                # below when comparing old state and corresponding new state.
+                # The exception is raised because there is no commit in change
+                # list corresponding to removal or renaming of END state.
+
+                # This problem arises when following scenario occurrs:
+                # Consider an exploration with two versions a and b (a < b).
+                # Both of this version has states dict schema version set to
+                # less than 2.
+                # Now version 'a' has explicit references to END
+                # state and thus when its states dict schema version will be
+                # upgraded to latest version, an 'END' state will be added.
+                # These explicit END references are not present in version 'b',
+                # however, and so no END state will be added during states dict
+                # schema migration.
+
+                # Thus we no commit of END state's removal and
```

Comment:
What exactly does 'no commit' refer to here?

Reply:
no commit as in we don't have any command in change list which says that END state has been removed or renamed.